c++/src/kj/compat/http.c++

```diff
@@ -5049,6 +5762,49 @@ private:
       }
 
       KJ_SWITCH_ONEOF(requestOrProtocolError) {
+        KJ_CASE_ONEOF(request, HttpHeaders::ConnectRequest) {
+          auto& headers = httpInput.getHeaders();
+
+          currentMethod = HttpConnectMethod();
+
+          SuspendableRequest suspendable(*this, HttpConnectMethod(), request.authority, headers);
+          auto maybeService = factory(suspendable);
+
+          if (suspended) {
+            return false;
+          }
+
+          auto service = KJ_ASSERT_NONNULL(kj::mv(maybeService),
+              "SuspendableHttpServiceFactory did not suspend, but returned nullptr.");
+
+          // TODO(perf): If the client disconnects, should we cancel the response? Probably, to
+          //   prevent permanent deadlock. It's slightly weird in that arguably the client should
+          //   be able to shutdown the upstream but still wait on the downstream, but I believe many
+          //   other HTTP servers do similar things.
+
+          // TODO(now): Perform any necessary validations before dispatching to the service.
+          auto promise = service->connect(request.authority, headers, *this).attach(kj::mv(service));
+          return promise.then([this]() mutable -> kj::Promise<bool> {
+            KJ_IF_MAYBE(p, tunnelRejected) {
+              // reject() was called to reject a CONNECT attempt.
+              // Finish sending and close the connection.
+              auto promise = kj::mv(*p);
+              tunnelRejected = nullptr;
+              return kj::mv(promise);
+            }
+
+            // TODO(now): Needed?
+            if (httpOutput.isBroken()) {
+              return false;
+            }
+
+            return httpOutput.flush().then([]() mutable -> kj::Promise<bool> {
+              // TODO(now): Determine if there is any reasonable path to reusing
```

Comment:
Certainly not if the connection was accepted.

If it was rejected, you could arguably accept other requests after sending the rejection. However, this may be asking for trouble if the client gets confused and thinks a connection was opened. I think we should make sure to send `Connection: close` on all rejections of CONNECT requests, and then always close the connection.